cli/tests/unit/timers_test.ts

```diff
@@ -346,6 +401,26 @@ Deno.test(async function timerMaxCpuBug() {
   assert(opsDispatched_ - opsDispatched < 10);
 });
 
+Deno.test(async function timerOrdering() {
+  const array: number[] = [];
+  setTimeout(() => {
+    array.push(1);
+    setTimeout(() => array.push(4));
+  }, 0);
+  setTimeout(() => {
+    array.push(2);
+    setTimeout(() => array.push(5));
+  }, 0);
+  setTimeout(() => {
+    array.push(3);
+    setTimeout(() => array.push(6));
+  }, 0);
+
+  await delay(100);
```

Comment:
This looks still a little bit flaky from my observation. ref: https://github.com/denoland/deno/runs/4418361252?check_suite_focus=true

How about creating a promise for each setTimeout call and waiting for all of them instead of delay(N)

Reply:
Seems like a genuine bug. A race would be missing [4,5,6] but it's only missing 6. Merits deeper investigation.

Reply:
According to the spec, a timer mustn't go off before other timers that were both scheduled before it and have a lesser or equal timeout. So the ordering of the numbers pushed into the array must hold, and the delay must expire after 3 is pushed, but the delay doesn't necessarily have to go off after any of the 4/5/6 timers do. And that's what I find when I test this – either [1,2,3], [1,2,3,4], [1,2,3,4,5] or [1,2,3,4,5,6], but not [1,2] or anything out of order.

Note that the spec doesn't say that a timeout of 0 must queue the action in the timer task queue immediately – step 5.3 in ["run steps after a timeout"](https://html.spec.whatwg.org/multipage/timers-and-user-prompts.html#run-steps-after-a-timeout) allows for an arbitrary additional wait. And it seems like an `op_sleep` with a timeout of 0 doesn't necessarily resolve at the next event loop turn. Should we queue the task immediately in that case?

In any case, I'll be taking @kt3k's suggestion.

Reply:
> Should we queue the task immediately in that case?

Come to think about it, no. This would let timers starve other parts of the event loop. Whereas with timers which are always backed by ops, the earliest that the tasks can be pushed to the timers queue is on the next event loop turn.

Reply:
I don't follow your line of reasoning.

We have timers A, B and C that expire at time T+0, who arm three more timers D, E, and F that expire at time T+1 (or T+4, doesn't matter), and a delay G that expires at time T+100.

Under normal conditions, they expire in order A to G because 0 < 1 < 100.

Under abnormal conditions, say an event loop stall or an OS scheduler fluke that pauses the program until T+200, then either A-G or A-C, G, D-F are reasonable behavior.

But A-E, G, F? No, never.

<hr>

A timer's expiry time is set to `now() + timeout` rather than `event_loop_tick_start_time() + timeout`. Put another way: timers have no shared time base.

Under normal conditions that doesn't matter because computers are fast: `now()` doesn't change meaningfully in the time it takes to execute an event loop "tick".

But if there is a stall of duration N between B and C, then there is an equivalent gap between E and F: E expires at time T+1 but F at time T+N.

Reply:
I don't know how tokio's timers are implemented –I have tried to follow the code and can't really make sense of it–, or how they interact with `FuturesUnordered` –which is probably more relevant for how they show up here. The spec requires that timers don't resolve before `now() + timeout` –which I'm assuming tokio does correctly–, and it also requires that a timer doesn't resolve before any timers that were scheduled before it and that have a lower or equal timeout. The code around `scheduledTimers` guarantees that that condition holds, but it doesn't guarantee anything else. And, purely in terms of the spec requirements, A-E, G, F is valid, because D-F are not required to be added to the timer task queue _the instant_ `setTimeout()` is called.

Reply:
For posterity: Bartek and I talked this through OOB.

IMO, it's a quality-of-implementation issue - expiry order should follow the Principle of Least Surprise, regardless of how relaxed the spec is - but it's pre-existing, not a regression, and therefore doesn't need to hold up this PR.